File: Annadata/app/src/main/java/com/mananhirak/annadata/activitys/TiffinModify.java
package com.mananhirak.annadata.activitys;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.text.HtmlCompat;

import android.app.Dialog;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;
import com.mananhirak.annadata.R;
import com.mananhirak.annadata.dbhelpers.DELIVERY_BOY;
import com.mananhirak.annadata.dbhelpers.MARKET;
import com.mananhirak.annadata.idtag.FireTag;
import com.mananhirak.annadata.idtag.Mtag;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TiffinModify extends AppCompatActivity {

    String TIFFIN_DOC_ID;
    TextView t1,t2,t3,t4,t5,t6,allReview;
    ImageButton reviewPhoto;
    Button acceptingOrder, notAcceptingOrder,setOrder,closeReview;
    FirebaseFirestore FIRE_db;
    FirebaseUser user;
    ImageView backButton,foodImage;
    EditText editText;
    ListenerRegistration listenerRegistration_of_orderLeft;
    Dialog dialog;
    ConstraintLayout constraintLayout;
    private void IdSaver() {

        editText =findViewById(R.id.act_tm_neworder);

        t1=findViewById(R.id.act_tm_foodname);
        t2=findViewById(R.id.act_tm_foodweight);
        t3=findViewById(R.id.act_tm_totalorder);
        t4=findViewById(R.id.act_tm_sellerfoodinfo);
        t5=findViewById(R.id.act_tm_deliverytime);
        t6=findViewById(R.id.act_tm_avg_rating);
        reviewPhoto=findViewById(R.id.act_tm_ratingreview);
        constraintLayout =findViewById(R.id.act_tm_layout_rating);

        acceptingOrder = findViewById(R.id.act_tm_conform);
        notAcceptingOrder = findViewById(R.id.act_tm_cancel);
        setOrder = findViewById(R.id.act_tm_set_order);

        backButton = findViewById(R.id.act_tm_back_button);

        foodImage = findViewById(R.id.act_tm_foodphoto);

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tiffin_modify);

        Intent intent=getIntent();
        TIFFIN_DOC_ID=intent.getStringExtra("TIFFIN_DOC_ID");

        FIRE_db=FirebaseFirestore.getInstance();
        user= FirebaseAuth.getInstance().getCurrentUser();

        IdSaver();

        backButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });

        FIRE_db.collection(FireTag.FireMarket).document(TIFFIN_DOC_ID)
                .get().addOnSuccessListener(this::setDELIVERY);

    }

    private void setDELIVERY(DocumentSnapshot documentSnapshot) {

        MARKET market = documentSnapshot.toObject(MARKET.class);
        DocumentReference documentSnapshotReference = documentSnapshot.getReference();
        assert market != null;
        String s = "Food : " + market.getFOOD_NAME();
        t1.setText(s);
        s = "Delivery Time : " + market.getFOOD_TIME();
        t5.setText(s);
        s = "No. Order  :  " + market.getLEFT_ORDER();
        OrderLeft(documentSnapshot);
        s = market.getFOOD_WEIGHT() + " KG";
        t2.setText(s);
        s="Food Info : "+market.getFOOD_INFO();
        t4.setText(s);

        SetAvgRating(market.getALL_RATING(),market.getALL_REVIEW());

        boolean isNoFoodLeft =market.getNO_FOOD_LEFT();

        if(isNoFoodLeft){
            notAcceptingOrder.setVisibility(View.INVISIBLE);
        }else{
            acceptingOrder.setVisibility(View.INVISIBLE);
        }

        if((market.getALL_REVIEW())==null){
            reviewPhoto.setVisibility(View.GONE);
        }

        reviewPhoto.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SeeReview(market.getALL_REVIEW());
            }
        });

        acceptingOrder.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(market.getLEFT_ORDER()!=0) {
                    documentSnapshotReference.update("no_FOOD_LEFT", false);
                    acceptingOrder.setVisibility(View.INVISIBLE);
                    notAcceptingOrder.setVisibility(View.VISIBLE);
                }else{
                    Toast.makeText(TiffinModify.this, "Please increase Food Order", Toast.LENGTH_SHORT).show();
                }
            }
        });

        notAcceptingOrder.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                documentSnapshotReference.update("no_FOOD_LEFT", true);
                acceptingOrder.setVisibility(View.VISIBLE);
                notAcceptingOrder.setVisibility(View.INVISIBLE);
            }
        });

        Glide.with(this)
                .load(market.getPHOTO_URL())
                .into(foodImage);

        setOrder.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                SetOrder(market.getTOTAL_FOOD(),market.getLEFT_ORDER(),documentSnapshotReference);
            }
        });

    }

    private void SeeReview(List<String> allRatingReview){



        dialog =new Dialog(this);
        dialog.setContentView(R.layout.dial_reviews);
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.create();
        dialog.show();

        allReview=dialog.findViewById(R.id.dial_r_allreview);

        StringBuilder review= new StringBuilder();

        for(String r : allRatingReview ){
            review.append(r).append(" \n ");
        }

        allReview.setText(review.toString());

        closeReview=dialog.findViewById(R.id.dial_r_close);

        closeReview.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dialog.cancel();
            }
        });

    }

    private void OrderLeft(DocumentSnapshot documentSnapshot) {
        listenerRegistration_of_orderLeft = documentSnapshot.getReference().addSnapshotListener((value, error) -> {
            if(error!=null){
                Toast.makeText(TiffinModify.this, error.getMessage(), Toast.LENGTH_SHORT).show();
            }else if(value==null){
                Toast.makeText(TiffinModify.this, "No Order Left", Toast.LENGTH_SHORT).show();
                finish();
            }else{
                Log.d(Mtag.Tag, "\n\n\n\nOrderLeft: "+value.get("left_ORDER")+"\n\n\n\n");
                String s=value.get("left_ORDER")+" Order Left";
                t3.setText(s);
            }

        });
    }

    private void SetAvgRating(List<Integer> rating,List<String> review){
        float avg=0;
        if(review==null){
            constraintLayout.setVisibility(View.INVISIBLE);
            return;
        }
        //TODO
        for (int rate:rating) {
            avg+=rate;
        }
        avg/=rating.size();
        String s =GetEmoji((int) Math.floor(avg))+" < Rating < "+GetEmoji((int)Math.ceil(avg));
        t6.setText(s);

    }

    private String GetEmoji(int value){
        switch (value){
            case 1:
                return "\uD83D\uDE1F";
            case 2:
                return "\uD83D\uDE41";
            case 3:
                return "\uD83D\uDE10";
            case 4:
                return "\uD83D\uDE42";
            default:
                return "\uD83D\uDE00";


        }
    }

    private  void SetOrder(int totalOrder,int leftOrder,DocumentReference documentReference){
       String newOrderString = editText.getText().toString();
        if(newOrderString.isEmpty()){
            editText.requestFocus();
            editText.setError("Please Enter new order number");
        }else{
           int newOrder = Integer.parseInt(newOrderString);
            Map<String,Object> map = new HashMap<>();
            map.put("left_ORDER",newOrder);
            map.put("total_FOOD",(totalOrder+newOrder-leftOrder));
            documentReference.update(map).addOnSuccessListener(new OnSuccessListener<Void>() {
               @Override
               public void onSuccess(Void unused) {
                   Toast.makeText(TiffinModify.this, "Order updated Successfully", Toast.LENGTH_SHORT).show();
               }
           })
            .addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    Toast.makeText(TiffinModify.this, "Order not updated!", Toast.LENGTH_SHORT).show();
                }
            });
           }

        }

    @Override
    protected void onStop() {
        super.onStop();
        if(listenerRegistration_of_orderLeft!=null)
            listenerRegistration_of_orderLeft.remove();
    }


}

File: Annadata/app/src/main/java/com/manan/annadata/activitys/TiffinModify.kt
package com.manan.annadata.activitys

import android.app.Dialog
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import com.bumptech.glide.Glide
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.*
import com.manan.annadata.R
import com.manan.annadata.dbhelpers.MARKET
import com.manan.annadata.idtag.FireTag
import com.manan.annadata.idtag.Mtag

class TiffinModify constructor() : AppCompatActivity() {
    var TIFFIN_DOC_ID: String? = null
    var t1: TextView? = null
    var t2: TextView? = null
    var t3: TextView? = null
    var t4: TextView? = null
    var t5: TextView? = null
    var t6: TextView? = null
    var allReview: TextView? = null
    var reviewPhoto: ImageButton? = null
    var acceptingOrder: Button? = null
    var notAcceptingOrder: Button? = null
    var setOrder: Button? = null
    var closeReview: Button? = null
    var FIRE_db: FirebaseFirestore? = null
    var user: FirebaseUser? = null
    var backButton: ImageView? = null
    var foodImage: ImageView? = null
    var editText: EditText? = null
    var listenerRegistration_of_orderLeft: ListenerRegistration? = null
    var dialog: Dialog? = null
    var constraintLayout: ConstraintLayout? = null
    private fun IdSaver() {
        editText = findViewById(R.id.act_tm_neworder)
        t1 = findViewById(R.id.act_tm_foodname)
        t2 = findViewById(R.id.act_tm_foodweight)
        t3 = findViewById(R.id.act_tm_totalorder)
        t4 = findViewById(R.id.act_tm_sellerfoodinfo)
        t5 = findViewById(R.id.act_tm_deliverytime)
        t6 = findViewById(R.id.act_tm_avg_rating)
        reviewPhoto = findViewById(R.id.act_tm_ratingreview)
        constraintLayout = findViewById(R.id.act_tm_layout_rating)
        acceptingOrder = findViewById(R.id.act_tm_conform)
        notAcceptingOrder = findViewById(R.id.act_tm_cancel)
        setOrder = findViewById(R.id.act_tm_set_order)
        backButton = findViewById(R.id.act_tm_back_button)
        foodImage = findViewById(R.id.act_tm_foodphoto)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tiffin_modify)
        val intent: Intent = getIntent()
        TIFFIN_DOC_ID = intent.getStringExtra("TIFFIN_DOC_ID")
        FIRE_db = FirebaseFirestore.getInstance()
        user = FirebaseAuth.getInstance().getCurrentUser()
        IdSaver()
        backButton!!.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(view: View) {
                finish()
            }
        })
        FIRE_db!!.collection(FireTag.FireMarket).document((TIFFIN_DOC_ID)!!)
            .get().addOnSuccessListener(OnSuccessListener({ documentSnapshot: DocumentSnapshot ->
                setDELIVERY(documentSnapshot)
            }))
    }

    private fun setDELIVERY(documentSnapshot: DocumentSnapshot) {
        val market: MARKET? = documentSnapshot.toObject(MARKET::class.java)
        val documentSnapshotReference: DocumentReference = documentSnapshot.getReference()
        assert(market != null)
        var s: String? =market!!.FOOD_NAME
        t1!!.setText(s)
        s = "Delivery Time : " + market.FOOD_TIME
        t5!!.setText(s)
        s = "No. Order  :  " + market.LEFT_ORDER
        OrderLeft(documentSnapshot)
        s = market.FOOD_WEIGHT.toString() + " KG"
        t2!!.setText(s)
        s = market.FOOD_INFO
        t4!!.setText(s)
        SetAvgRating(market.ALL_RATING, market.ALL_REVIEW)
        val isNoFoodLeft: Boolean = market.NO_FOOD_LEFT
        if (isNoFoodLeft) {
            notAcceptingOrder!!.setVisibility(View.INVISIBLE)
        } else {
            acceptingOrder!!.setVisibility(View.INVISIBLE)
        }
        if ((market.ALL_REVIEW) == null) {
            reviewPhoto!!.setVisibility(View.GONE)
        }
        reviewPhoto!!.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(view: View) {
                SeeReview(market.ALL_REVIEW)
            }
        })
        acceptingOrder!!.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(view: View) {
                if (market.LEFT_ORDER != 0) {
                    documentSnapshotReference.update("no_FOOD_LEFT", false)
                    acceptingOrder!!.setVisibility(View.INVISIBLE)
                    notAcceptingOrder!!.setVisibility(View.VISIBLE)
                } else {
                    Toast.makeText(
                        this@TiffinModify,
                        "Please increase Food Order",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
        })
        notAcceptingOrder!!.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(view: View) {
                documentSnapshotReference.update("no_FOOD_LEFT", true)
                acceptingOrder!!.setVisibility(View.VISIBLE)
                notAcceptingOrder!!.setVisibility(View.INVISIBLE)
            }
        })
        Glide.with(this)
            .load(market.PHOTO_URL)
            .into((foodImage)!!)
        setOrder!!.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(view: View) {
                SetOrder(market.TOTAL_FOOD, market.LEFT_ORDER, documentSnapshotReference)
            }
        })
    }

    private fun SeeReview(allRatingReview: List<String>?) {
        dialog = Dialog(this)
        dialog!!.setContentView(R.layout.dial_reviews)
        dialog!!.getWindow()!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog!!.create()
        dialog!!.show()
        allReview = dialog!!.findViewById(R.id.dial_r_allreview)
        val review: StringBuilder = StringBuilder()
        for (r: String in allRatingReview!!) {
            review.append(r).append(" \n ")
        }
        allReview!!.setText(review.toString())
        closeReview = dialog!!.findViewById(R.id.dial_r_close)
        closeReview!!.setOnClickListener(object : View.OnClickListener {
            public override fun onClick(view: View) {
                dialog!!.cancel()
            }
        })
    }

    private fun OrderLeft(documentSnapshot: DocumentSnapshot) {
        listenerRegistration_of_orderLeft = documentSnapshot.getReference().addSnapshotListener(
            EventListener({ value: DocumentSnapshot?, error: FirebaseFirestoreException? ->
                if (error != null) {
                    Toast.makeText(this@TiffinModify, error.message, Toast.LENGTH_SHORT).show()
                } else if (value == null) {
                    Toast.makeText(this@TiffinModify, "No Order Left", Toast.LENGTH_SHORT).show()
                    finish()
                } else {
                    Log.d(Mtag.Tag, "\n\n\n\nOrderLeft: " + value.get("left_ORDER") + "\n\n\n\n")
                    val s: String = value.get("left_ORDER").toString() + " Order Left"
                    t3!!.setText(s)
                }
            })
        )
    }

    private fun SetAvgRating(rating: List<Int>?, review: List<String>?) {
        var avg: Float = 0f
        if (review == null) {
            constraintLayout!!.setVisibility(View.INVISIBLE)
            return
        }
        //TODO
        for (rate: Int in rating!!) {
            avg += rate.toFloat()
        }
        avg /= rating.size.toFloat()
        val s: String = GetEmoji(Math.floor(avg.toDouble()).toInt()) + " < Rating < " + GetEmoji(
            Math.ceil(avg.toDouble()).toInt()
        )
        t6!!.setText(s)
    }

    private fun GetEmoji(value: Int): String {
        when (value) {
            1 -> return "\uD83D\uDE1F"
            2 -> return "\uD83D\uDE41"
            3 -> return "\uD83D\uDE10"
            4 -> return "\uD83D\uDE42"
            else -> return "\uD83D\uDE00"
        }
    }

    private fun SetOrder(totalOrder: Int, leftOrder: Int, documentReference: DocumentReference) {
        val newOrderString: String = editText!!.getText().toString()
        if (newOrderString.isEmpty()) {
            editText!!.requestFocus()
            editText!!.setError("Please Enter new order number")
        } else {
            val newOrder: Int = newOrderString.toInt()
            val map: MutableMap<String, Any> = HashMap()
            map.put("left_ORDER", newOrder)
            map.put("total_FOOD", (totalOrder + newOrder - leftOrder))
            documentReference.update(map).addOnSuccessListener(object : OnSuccessListener<Void?> {
                public override fun onSuccess(unused: Void?) {
                    Toast.makeText(
                        this@TiffinModify,
                        "Order updated Successfully",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            })
                .addOnFailureListener(object : OnFailureListener {
                    public override fun onFailure(e: Exception) {
                        Toast.makeText(this@TiffinModify, "Order not updated!", Toast.LENGTH_SHORT)
                            .show()
                    }
                })
        }
    }

    override fun onStop() {
        super.onStop()
        if (listenerRegistration_of_orderLeft != null) listenerRegistration_of_orderLeft!!.remove()
    }
}